tsl/src/nodes/vector_agg/plan_decompress_chunk.c

```diff
@@ -0,0 +1,139 @@
+/*
+ * This file and its contents are licensed under the Timescale License.
+ * Please see the included NOTICE for copyright information and
+ * LICENSE-TIMESCALE for a copy of the license.
+ */
+#include <postgres.h>
+#include <nodes/pathnodes.h>
+#include <nodes/plannodes.h>
+
+#include "nodes/decompress_chunk/planner.h"
+#include "plan.h"
+
+/*
+ * Whether the given compressed column index corresponds to a vector variable.
+ */
+static bool
+is_vector_compressed_column(const CustomScan *custom, int compressed_column_index,
+							bool *out_is_segmentby)
+{
+	List *bulk_decompression_column = list_nth(custom->custom_private, DCP_BulkDecompressionColumn);
+	const bool bulk_decompression_enabled_for_column =
+		list_nth_int(bulk_decompression_column, compressed_column_index);
+
+	/*
+	 * Bulk decompression can be disabled for all columns in the DecompressChunk
+	 * node settings, we can't do vectorized aggregation for compressed columns
+	 * in that case. For segmentby columns it's still possible.
+	 */
+	List *settings = linitial(custom->custom_private);
+	const bool bulk_decompression_enabled_globally =
+		list_nth_int(settings, DCS_EnableBulkDecompression);
+
+	/*
+	 * Check if this column is a segmentby.
+	 */
+	List *is_segmentby_column = list_nth(custom->custom_private, DCP_IsSegmentbyColumn);
+	const bool is_segmentby = list_nth_int(is_segmentby_column, compressed_column_index);
+	if (out_is_segmentby)
+	{
+		*out_is_segmentby = is_segmentby;
+	}
+
+	/*
+	 * We support vectorized aggregation either for segmentby columns or for
+	 * columns with bulk decompression enabled.
+	 */
+	if (!is_segmentby &&
+		!(bulk_decompression_enabled_for_column && bulk_decompression_enabled_globally))
+	{
+		/* Vectorized aggregation not possible for this particular column. */
+		return false;
+	}
+
+	return true;
+}
+
+/*
+ * Map the custom scan attribute number to the uncompressed chunk attribute
+ * number.
+ */
+static int
+custom_scan_to_uncompressed_chunk_attno(List *custom_scan_tlist, int custom_scan_attno)
+{
+	if (custom_scan_tlist == NIL)
+	{
+		return custom_scan_attno;
+	}
+
+	Var *var =
+		castNode(Var,
+				 castNode(TargetEntry,
+						  list_nth(custom_scan_tlist, AttrNumberGetAttrOffset(custom_scan_attno)))
+					 ->expr);
```

Comment:
Nit: it is easier to read if each `castNode` is assigned to a separate variable.